mono/metadata/threads.c

```diff
@@ -691,6 +691,8 @@ mono_thread_attach_internal (MonoThread *thread, gboolean force_attach, gboolean
 		return FALSE;
 	}
 
+	mono_thread_info_set_detach_user_data (info, (gpointer) internal);
```

Comment:
I am not very happy with this solution, so I am exploring others. If we decide to stick with this one, I will have to register a GC root in MonoThreadInfo.